rustler_codegen/src/init.rs

```diff
@@ -0,0 +1,111 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::parse::{Parse, ParseStream};
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+use syn::{Expr, Result, Token};
+
+#[derive(Debug)]
+pub struct InitMacroInput {
+    module: syn::Lit,
+    funcs: syn::ExprArray,
+    on_load: Option<TokenStream>,
+}
+
+impl Parse for InitMacroInput {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let module = syn::Lit::parse(input)?;
+        let _comma = <syn::Token![,]>::parse(input)?;
+        let funcs = syn::ExprArray::parse(input)?;
+        let on_load = if input.peek(Token![,]) {
+            let _comma = <Token![,]>::parse(input)?;
+            Some(TokenStream::parse(input)?)
+        } else {
+            None
+        };
+
+        Ok(InitMacroInput {
+            module,
+            funcs,
+            on_load,
+        })
+    }
+}
+
+impl Into<proc_macro2::TokenStream> for InitMacroInput {
+    fn into(self) -> proc_macro2::TokenStream {
+        let name = self.module;
+        let num_of_funcs = self.funcs.elems.len();
+        let funcs = nif_funcs(self.funcs.elems);
+        let on_load = self.on_load;
+
+        let inner = quote! {
+            static mut NIF_ENTRY: Option<rustler::codegen_runtime::DEF_NIF_ENTRY> = None;
+            use rustler::Nif;
+
+            let entry = rustler::codegen_runtime::DEF_NIF_ENTRY {
+                major: rustler::codegen_runtime::NIF_MAJOR_VERSION,
+                minor: rustler::codegen_runtime::NIF_MINOR_VERSION,
+                name: concat!(#name, "\0").as_ptr() as *const u8,
+                num_of_funcs: #num_of_funcs as rustler::codegen_runtime::c_int,
+                funcs: [#funcs].as_ptr(),
+                load: {
+                    extern "C" fn nif_load(
+                        env: rustler::codegen_runtime::NIF_ENV,
+                        _priv_data: *mut *mut rustler::codegen_runtime::c_void,
+                        load_info: rustler::codegen_runtime::NIF_TERM
+                    ) -> rustler::codegen_runtime::c_int {
+                        unsafe {
+                            // TODO: If an unwrap ever happens, we will unwind right into C! Fix this!
+                            rustler::codegen_runtime::handle_nif_init_call(#on_load, env, load_info)
```

Comment:
WRT https://github.com/rusterlium/rustler/pull/264#issuecomment-548420083:

When `self.on_load` is `None` (https://github.com/rusterlium/rustler/pull/264/files#diff-8daf9e0986d966ebc2c4878e8a772b5fR40), this will not be expanded. It seems that we must pass-in a Tokenstream here, even if we do not have an `on_load` function.

Reply:
We could parse `on_load` as follows, which would make this work here:

```rust
        let on_load = if input.peek(Token![,]) {
            let _comma = <Token![,]>::parse(input)?;
            TokenStream::parse(input)?
        } else {
            let none_ident = Ident::new("None", Span::call_site());
            Some(quote!{#none_ident})
        };
```

We should then probably change the type of `on_load` from `Option<TokenStream>` to `TokenStream`, as it will be `Some(..)` anyways.